src/mscorlib/corefx/System/Globalization/CompareInfo.Unix.cs

```diff
@@ -165,75 +188,30 @@ private bool EndsWith(string source, string suffix, CompareOptions options)
             Contract.Assert(!string.IsNullOrEmpty(suffix));
             Contract.Assert((options & (CompareOptions.Ordinal | CompareOptions.OrdinalIgnoreCase)) == 0);
 
-            // TODO: Implement This Fully.
-            if(suffix.Length > source.Length) return false;
-
-            return StringEqualsAscii(source.Substring(source.Length - suffix.Length), suffix, IgnoreCase(options));
+            return Interop.GlobalizationInterop.EndsWith(m_sortNameAsUtf8, suffix, source, source.Length, options);
         }
 
         // -----------------------------
         // ---- PAL layer ends here ----
         // -----------------------------
 
-        private static char ChangeCaseAscii(char c, bool toUpper = true)
+        internal unsafe int GetHashCodeOfStringCore(string source, CompareOptions options, bool forceRandomizedHashing, long additionalEntropy)
         {
-            if (toUpper && c >= 'a' && c <= 'z')
-            {
-                return (char)('A' + (c - 'a'));
-            }
-            else if (!toUpper && c >= 'A' && c <= 'Z')
-            {
-                return (char)('a' + (c - 'A'));
-            }
-
-            return c;
-        }
-
-        private static bool StringEqualsAscii(string s1, string s2, bool ignoreCase = true)
-        {
-            if (s1.Length != s2.Length) return false;
-
-            for (int i = 0; i < s1.Length; i++)
-            {
-                char c1 = ignoreCase ? ChangeCaseAscii(s1[i]) : s1[i];
-                char c2 = ignoreCase ? ChangeCaseAscii(s2[i]) : s2[i];
+            Contract.Assert(source != null);
+            Contract.Assert((options & (CompareOptions.Ordinal | CompareOptions.OrdinalIgnoreCase)) == 0);
 
-                if (c1 != c2) return false;
-            }
+            int sortKeyLength = Interop.GlobalizationInterop.GetSortKey(m_sortNameAsUtf8, source, source.Length, null, 0, options);
 
-            return true;
-        }
+            byte[] sortKey = new byte[sortKeyLength];
```

Comment:
One possible future optimization would be to restructure this so that the shim does the sort key generation and hashing in tandem.  This would require us to pull Marvin32 into System.Globalization.Native.so (which may not be that bad). If we did that we could iterate across elements of the sort key without having to pre-allocate the sort key length up front.